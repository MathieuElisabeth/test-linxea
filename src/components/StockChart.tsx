import { useMemo } from 'react';
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer } from 'recharts';
import type { StockEvolution } from '@/types/stock';
import { formatCurrency, formatPercentage } from '@utils/stockUtils';
import CustomTooltip from './CustomTooltip';

interface StockChartProps {
  stockEvolution: StockEvolution | null;
}

interface ChartData {
  date: string;
  close: number;
  year: string;
}

export const StockChart = ({ stockEvolution }: StockChartProps) => {
  const chartData = useMemo((): ChartData[] => {
    if (!stockEvolution) return [];
    
    return stockEvolution.data.map(item => ({
      date: item.date,
      close: item.close,
      year: new Date(item.date).getFullYear().toString()
    }));
  }, [stockEvolution]);

  if (!stockEvolution) {
    return (
      <div className="chart-container">
        <div className="no-data">
          Sélectionnez une action pour voir son évolution
        </div>
      </div>
    );
  }



  return (
    <div className="chart-container">
      <div className="chart-header">
        <h2>{stockEvolution.ticker}</h2>
        <div className={`evolution ${stockEvolution.evolution >= 0 ? 'positive' : 'negative'}`}>
          {formatPercentage(stockEvolution.evolution)}
        </div>
      </div>
      
      <ResponsiveContainer width="100%" height={400}>
        <LineChart data={chartData}>
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis 
            dataKey="year" 
            tick={{ fontSize: 12 }}
            tickLine={false}
          />
          <YAxis 
            tickFormatter={(value) => formatCurrency(value)}
            tick={{ fontSize: 12 }}
            tickLine={false}
          />
          <Tooltip content={<CustomTooltip />} />
          <Line 
            type="monotone" 
            dataKey="close" 
            stroke="#8884d8" 
            strokeWidth={3}
            dot={{ fill: '#8884d8', strokeWidth: 2, r: 4 }}
            activeDot={{ r: 6 }}
          />
        </LineChart>
      </ResponsiveContainer>
      
      <div className="chart-info">
        <div className="info-item">
          <span className="label">Prix initial:</span>
          <span className="value">{formatCurrency(chartData[0]?.close || 0)}</span>
        </div>
        <div className="info-item">
          <span className="label">Prix final:</span>
          <span className="value">{formatCurrency(chartData[chartData.length - 1]?.close || 0)}</span>
        </div>
        <div className="info-item">
          <span className="label">Évolution:</span>
          <span className={`value ${stockEvolution.evolution >= 0 ? 'positive' : 'negative'}`}>
            {formatPercentage(stockEvolution.evolution)}
          </span>
        </div>
      </div>
    </div>
  );
}; 