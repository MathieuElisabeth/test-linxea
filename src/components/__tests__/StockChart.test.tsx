import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { StockChart } from '../StockChart';
import type { StockEvolution } from '@/types/stock';

vi.mock('recharts', () => ({
  LineChart: ({ children }: { children: React.ReactNode }) => <div data-testid="line-chart">{children}</div>,
  Line: () => <div data-testid="line" />,
  XAxis: () => <div data-testid="x-axis" />,
  YAxis: () => <div data-testid="y-axis" />,
  CartesianGrid: () => <div data-testid="cartesian-grid" />,
  Tooltip: () => <div data-testid="tooltip" />,
  ResponsiveContainer: ({ children }: { children: React.ReactNode }) => <div data-testid="responsive-container">{children}</div>,
}));

describe('StockChart', () => {
  const mockStockEvolution: StockEvolution = {
    ticker: 'AAPL',
    data: [
      { date: '2020-12-31', ticker: 'AAPL', close: 132.69, volume: 99116671 },
      { date: '2021-12-31', ticker: 'AAPL', close: 177.57, volume: 74537652 },
      { date: '2022-12-30', ticker: 'AAPL', close: 129.93, volume: 91420614 },
      { date: '2023-12-29', ticker: 'AAPL', close: 182.09, volume: 68341205 },
      { date: '2024-12-31', ticker: 'AAPL', close: 200.15, volume: 70512345 },
    ],
    evolution: 50.85,
  };

  it('renders correctly when no stock evolution is provided', () => {
    const { container } = render(<StockChart stockEvolution={null} />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with stock evolution data', () => {
    const { container } = render(<StockChart stockEvolution={mockStockEvolution} />);
    expect(container).toMatchSnapshot();
  });

  it('displays correct stock information when data is provided', () => {
    render(<StockChart stockEvolution={mockStockEvolution} />);
    
    expect(screen.getByText('AAPL')).toBeInTheDocument();
    
    const percentageElements = screen.getAllByText('+50.85%');
    expect(percentageElements.length).toBeGreaterThan(0);
    
    expect(screen.getByText('$132.69')).toBeInTheDocument();
    expect(screen.getByText('$200.15')).toBeInTheDocument();
    
    expect(screen.getByText('Prix initial:')).toBeInTheDocument();
    expect(screen.getByText('Prix final:')).toBeInTheDocument();
    expect(screen.getByText('Ã‰volution:')).toBeInTheDocument();
  });
}); 