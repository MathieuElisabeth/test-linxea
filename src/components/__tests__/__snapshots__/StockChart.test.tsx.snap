// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StockChart > renders correctly when no stock evolution is provided 1`] = `
<div>
  <div
    class="chart-container"
  >
    <div
      class="no-data"
    >
      Sélectionnez une action pour voir son évolution
    </div>
  </div>
</div>
`;

exports[`StockChart > renders correctly with stock evolution data 1`] = `
<div>
  <div
    class="chart-container"
  >
    <div
      class="chart-header"
    >
      <h2>
        AAPL
      </h2>
      <div
        class="evolution positive"
      >
        +50.85%
      </div>
    </div>
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="cartesian-grid"
        />
        <div
          data-testid="x-axis"
        />
        <div
          data-testid="y-axis"
        />
        <div
          data-testid="tooltip"
        />
        <div
          data-testid="line"
        />
      </div>
    </div>
    <div
      class="chart-info"
    >
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Prix initial:
        </span>
        <span
          class="value"
        >
          $132.69
        </span>
      </div>
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Prix final:
        </span>
        <span
          class="value"
        >
          $200.15
        </span>
      </div>
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Évolution:
        </span>
        <span
          class="value positive"
        >
          +50.85%
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`StockChart > renders correctly with zero evolution 1`] = `
<div>
  <div
    class="chart-container"
  >
    <div
      class="chart-header"
    >
      <h2>
        AAPL
      </h2>
      <div
        class="evolution positive"
      >
        +0.00%
      </div>
    </div>
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="cartesian-grid"
        />
        <div
          data-testid="x-axis"
        />
        <div
          data-testid="y-axis"
        />
        <div
          data-testid="tooltip"
        />
        <div
          data-testid="line"
        />
      </div>
    </div>
    <div
      class="chart-info"
    >
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Prix initial:
        </span>
        <span
          class="value"
        >
          $132.69
        </span>
      </div>
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Prix final:
        </span>
        <span
          class="value"
        >
          $200.15
        </span>
      </div>
      <div
        class="info-item"
      >
        <span
          class="label"
        >
          Évolution:
        </span>
        <span
          class="value positive"
        >
          +0.00%
        </span>
      </div>
    </div>
  </div>
</div>
`;
